src/io/flutter/logging/tree/DataPanel.java

```diff
@@ -0,0 +1,282 @@
+  }
+
+  private void linkSelected(URL url) {
+    // TODO(pq): handle links
+  }
+
+  public void update(Object data) {
+    // Remove stale content.
+    for (Component c : getComponents()) {
+      remove(c);
+    }
+
+    boolean showDataPane = false;
+    if (data instanceof DiagnosticsNode) {
+      showDataPane = updateNodeData((DiagnosticsNode)data);
+    }
+    else if (data instanceof String && JsonUtils.hasJsonData((String)data)) {
+      showDataPane = updateJsonTextData((String)data);
+    }
+
+    showPane(showDataPane);
+  }
+
+  private void showPane(boolean show) {
+    setVisible(show);
+    if (show) {
+      repaint();
+      revalidate();
+    }
+    dispatcher.getMulticaster().hasContent(show);
+
+    // TODO(pq): set caret position
+    // Scroll to start.
+    // setCaretPosition(0);
+  }
+
+  private boolean updateJsonTextData(String data) {
+    @SuppressWarnings("ConstantConditions") final JsonElement jsonElement = new JsonParser().parse((String)data);
+    final String text = gsonHelper.toJson(jsonElement);
+    if (text.isEmpty()) {
+      return false;
+    }
+
+    final JEditorPane editorPane = createEditorPane();
+    editorPane.setText(html(pre(text)));
+    add(editorPane);
+
+    return true;
+  }
+
+  private boolean updateNodeData(DiagnosticsNode data) {
+    // Header.
+    addText(data.toString(), data);
+
+    // Body.
+    final List<DiagnosticsNode> properties = data.getInlineProperties();
+    for (DiagnosticsNode node : properties) {
+      String contents = "";
+      if (node.getName() != null) {
+        contents += node.getName();
+      }
+
+      final ArrayList<DiagnosticsNode> children = getChildren(node);
+      if (!children.isEmpty() || node.getDescription() != null) {
+        if (!contents.isEmpty()) {
+          contents += ": ";
+        }
+        if (node.getDescription() != null) {
+          contents += node.getDescription();
+        }
+      }
+
+      if (children.isEmpty()) {
+        addText(contents, node);
+      }
+      else {
+        addTree(contents, node);
+      }
+    }
+    return true;
+  }
+
+  private void addText(String contents, DiagnosticsNode node) {
+    final JEditorPane editorPane = createEditorPane();
+    editorPane.setText(html(
+      div(
+        CssStyle.forNode(node),
+        span(contents)
+      )));
+    add(editorPane);
+  }
+
+  private void addTree(String rootLabel, DiagnosticsNode node) {
+    final JTree tree = new JTree();
+    tree.setModel(new NodeModel(node, rootLabel));
+    tree.setCellRenderer(new NodeRenderer());
+    tree.setShowsRootHandles(true);
+    tree.collapseRow(0);
+
+    tree.addFocusListener(new FocusAdapter() {
+      @Override
+      public void focusLost(FocusEvent e) {
+        tree.clearSelection();
+      }
+    });
+
+    final NonOpaquePanel panel = new NonOpaquePanel();
+    panel.add(tree);
+    add(panel);
+  }
+
+  private JEditorPane createEditorPane() {
+    final JEditorPane editorPane = new JEditorPane();
```

Comment:
Interesting use of editor pane.

Reply:
Yeah.  It handles HTML and reflows pretty well.